File: Competitive Programming/Blind 75 Must Do Leetcode/Top K Frequent Elements.py
'''https://leetcode.com/problems/top-k-frequent-elements/
347. Top K Frequent Elements
Medium

6346

304

Add to List

Share
Given an integer array nums and an integer k, return the k most frequent elements. You may return the answer in any order.

 

Example 1:

Input: nums = [1,1,1,2,2,3], k = 2
Output: [1,2]
Example 2:

Input: nums = [1], k = 1
Output: [1]
 

Constraints:

1 <= nums.length <= 105
k is in the range [1, the number of unique elements in the array].
It is guaranteed that the answer is unique.
 

Follow up: Your algorithm's time complexity must be better than O(n log n), where n is the array's size.'''

# Time:  O(n)
# Space: O(n)

# Given a non-empty array of integers,
# return the k most frequent elements.
#
# For example,
# Given [1,1,1,2,2,3] and k = 2, return [1,2].
#
# Note:
# You may assume k is always valid,
# 1 <= k <= number of unique elements.
# Your algorithm's time complexity must be better
# than O(n log n), where n is the array's size.

# Bucket Sort Solution

from random import randint
import collections

try:
    range          # Python 2
except NameError:
    range = range  # Python 3


class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        counts = collections.Counter(nums)
        buckets = [[] for _ in range(len(nums)+1)]
        for i, count in counts.iteritems():
            buckets[count].append(i)

        result = []
        for i in reversed(range(len(buckets))):
            for j in range(len(buckets[i])):
                result.append(buckets[i][j])
                if len(result) == k:
                    return result
        return result


# Time:  O(n) ~ O(n^2), O(n) on average.
# Space: O(n)
# Quick Select Solution


class Solution2(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        counts = collections.Counter(nums)
        p = []
        for key, val in counts.iteritems():
            p.append((-val, key))
        self.kthElement(p, k)

        result = []
        for i in range(k):
            result.append(p[i][1])
        return result

    def kthElement(self, nums, k):
        def PartitionAroundPivot(left, right, pivot_idx, nums):
            pivot_value = nums[pivot_idx]
            new_pivot_idx = left
            nums[pivot_idx], nums[right] = nums[right], nums[pivot_idx]
            for i in range(left, right):
                if nums[i] < pivot_value:
                    nums[i], nums[new_pivot_idx] = nums[new_pivot_idx], nums[i]
                    new_pivot_idx += 1

            nums[right], nums[new_pivot_idx] = nums[new_pivot_idx], nums[right]
            return new_pivot_idx

        left, right = 0, len(nums) - 1
        while left <= right:
            pivot_idx = randint(left, right)
            new_pivot_idx = PartitionAroundPivot(left, right, pivot_idx, nums)
            if new_pivot_idx == k - 1:
                return
            elif new_pivot_idx > k - 1:
                right = new_pivot_idx - 1
            else:  # new_pivot_idx < k - 1.
                left = new_pivot_idx + 1


# Time:  O(nlogk)
# Space: O(n)
class Solution3(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        return [key for key, _ in collections.Counter(nums).most_common(k)]


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        res = []
        dict = collections.Counter(nums)
        for val, count in dict.items():
            if len(res) < k:
                heapq.heappush(res, (count, val))
            else:
                heapq.heappush(res, (count, val))
                heapq.heappop(res)
        return [val for count, val in res]

File: competitive-programming/Blind 75 Must Do Leetcode/Top K Frequent Elements.cpp
/*
https://leetcode.com/problems/top-k-frequent-elements/
347. Top K Frequent Elements
Medium

6346

304

Add to List

Share
Given an integer array nums and an integer k, return the k most frequent elements. You may return the answer in any order.

 

Example 1:

Input: nums = [1,1,1,2,2,3], k = 2
Output: [1,2]
Example 2:

Input: nums = [1], k = 1
Output: [1]
 

Constraints:

1 <= nums.length <= 105
k is in the range [1, the number of unique elements in the array].
It is guaranteed that the answer is unique.
 

Follow up: 
Your algorithm's time complexity must be better than O(n log n), where n is the array's size.
*/

//Runtime: 24 ms, faster than 39.54% of C++ online submissions for Top K Frequent Elements.
//Memory Usage: 10.7 MB, less than 100.00% of C++ online submissions for Top K Frequent Elements.

class valuecomparator
{
public:
    valuecomparator() {}
    bool operator()(const pair<int, int> &lhs, const pair<int, int> &rhs) const
    {
        return lhs.second < rhs.second;
    }
};

class Solution
{
public:
    vector<int> topKFrequent(vector<int> &nums, int k)
    {
        map<int, int> counter;
        vector<int> ans;

        for (int e : nums)
        {
            counter[e]++;
        }

        //cannot sort map!
        // sort(counter.begin(), counter.end(),
        //      [](const pair<int, int>& a, const pair<int, int>& b){
        //          return a.second < b.second;
        // });

        priority_queue<pair<int, int>, vector<pair<int, int> >, valuecomparator> pq1;

        for (auto it = counter.begin(); it != counter.end(); it++)
        {
            pq1.push(*it);
        }

        for (int i = 0; i < k; i++)
        {
            pair<int, int> p = pq1.top();
            ans.push_back(p.first);
            pq1.pop();
        }

        return ans;
    }
};

//official solution
//improve efficiency by keeping priority queue small
//Runtime: 24 ms, faster than 39.54% of C++ online submissions for Top K Frequent Elements.
//Memory Usage: 10.3 MB, less than 100.00% of C++ online submissions for Top K Frequent Elements.
class valuecomparator
{
public:
    valuecomparator() {}
    bool operator()(const pair<int, int> &lhs, const pair<int, int> &rhs) const
    {
        return lhs.second > rhs.second;
    }
};

class Solution
{
public:
    vector<int> topKFrequent(vector<int> &nums, int k)
    {
        map<int, int> counter;
        vector<int> ans(k);

        for (int e : nums)
        {
            counter[e]++;
        }

        //the larger the lower
        priority_queue<pair<int, int>, vector<pair<int, int> >, valuecomparator> pq1;

        for (auto it = counter.begin(); it != counter.end(); it++)
        {
            pq1.push(*it);
            //the smallest will be popped first
            if (pq1.size() > k)
                pq1.pop();
        }

        //put into ans in reverse order
        for (int i = k - 1; i >= 0; i--)
        {
            pair<int, int> p = pq1.top();
            ans[i] = p.first;
            pq1.pop();
        }

        return ans;
    }
};